diesel/src/query_source/mod.rs

```diff
@@ -76,3 +76,83 @@ pub trait AppearsInFromClause<QS> {
     /// How many times does `Self` appear in `QS`?
     type Count;
 }
+
+#[doc(hidden)]
+/// Used to determine which of two from clauses contains a given table.
+///
+/// This trait can be used to emulate "or" conditions in where clauses when
+/// we want a trait to be implemented with one of two type parameters.
+///
+/// For example, if we wanted to write:
+///
+/// ```rust,ignore
+/// where
+///     T: SelectableExpression<Left> | SelectableExpression<Right>,
+/// ```
+///
+/// we can emulate this by writing:
+///
+/// ```rust,ignore
+/// where
+///     Left: AppearsInFromClause<T::Table>,
+///     Right: AppearsInFromClause<T::Table>,
+///     (Left::Count, Right::Count): Select<Left, Right>,
+///     T: SelectableExpression<
+///         <(Left::Count, Right::Count) as Select<Left, Right>>::Selection,
+///     >,
+/// ```
+///
+/// In order to aquire the counts in the first place, we must already know
+/// the table we're searching for.
+pub trait Select<Left, Right> {
```

Comment:
I would vote for another name, because `Select` collides with select from SQL statements.
`MatchByOnce`?

Reply:
Renamed to `Pick`